d3dcore/include/d3dcore/d3d_info.h

```diff
@@ -0,0 +1,253 @@
+#pragma once
+
+#define INITGUID
+#include "dxdiag.h"
+#include <Windows.h>
+
+#include "d3d_utils.h"
+
+#include <string>
+#include <memory>
+#include <assert.h>
+#include <iostream>
+#include <string>
+#include <any>
+
+
+//-----------------------------------------------------------------------------
+// Defines, and constants
+//-----------------------------------------------------------------------------
+#define SAFE_DELETE(p)       { if(p) { delete (p);     (p)=NULL; } }
+#define SAFE_DELETE_ARRAY(p) { if(p) { delete[] (p);   (p)=NULL; } }
+#define SAFE_RELEASE(p)      { if(p) { (p)->Release(); (p)=NULL; } }
+#define SAFE_BSTR_FREE(x)    if(x) { SysFreeString( x ); x = NULL; }
+#define EXPAND(x)            x, sizeof(x)/sizeof(TCHAR)
+
+
+
+namespace d3dcore {
+
+	/*enum class SDKVersion {
+	DX9: DXDIAG_DX9_SDK_VERSION,
+	DX11: 112,		
+	};*/
+
+	template<typename Container>
+	class D3DInfo
+	{
+	public:
+		D3DInfo();
+		~D3DInfo();
+
+
+	private:
+		//property
+		IDxDiagProvider* m_dxdiag_provider;
+		IDxDiagContainer* m_dxdiag_root;
+		std::shared_ptr<Container> m_root_container;
+		//private method
+		std::string get_property_value(IDxDiagContainer* dxContainer, WCHAR* propName); //todo: update with sdt::any
+		
+		
+	public:
+		void init(int dxVersion = DXDIAG_DX9_SDK_VERSION);
+		void traverse(IDxDiagContainer*, std::shared_ptr<Container> parent);
+		void query_info();
+		std::shared_ptr<Container> get_root_container();
+	};
+
+}
+
+
+using namespace d3dcore;
+
+template<typename Container>
+D3DInfo<Container>::D3DInfo()
+{
+	m_dxdiag_provider = nullptr;
+	m_dxdiag_root = nullptr;
+	m_root_container = std::make_shared<Container>("DirectX");
+}
+
+template<typename Container>
+D3DInfo<Container>::~D3DInfo()
+{
+	SAFE_RELEASE(m_dxdiag_provider);
+	SAFE_RELEASE(m_dxdiag_root);
+}
+
+template<typename Container>
+void D3DInfo<Container>::query_info()
+{
+
+	traverse(m_dxdiag_root, m_root_container);
+
+
+}
+
+template<typename Container>
+std::shared_ptr<Container> D3DInfo<Container>::get_root_container()
+{
+	return this->m_root_container;
+}
+
+//}
+
+
+
+
+template<typename Container>
+std::string D3DInfo<Container>::get_property_value(IDxDiagContainer * dxContainer, WCHAR * propName)
+{
+	HRESULT hr;
+	VARIANT var;
+	VariantInit(&var);
+	std::string result;
+
+	if (FAILED(hr = dxContainer->GetProp(propName, &var))) return std::string();
+
+
+	switch (var.vt) {
+
+	case VT_I4:
+		result = std::to_string(var.intVal);
+		break;
+
+	case VT_UI4:
+		result = std::to_string(var.uintVal);
+		break;
+
+	case VT_BOOL:
+		result = var.boolVal != 0 ? std::string("true") : std::string("false");
+		break;
+
+		/*case VT_BSTR:
+
+	#ifdef _UNICODE
+			wcsncpy(strValue, var.bstrVal, 256); //todo: update the length of string
+	#else
+			wcstombs(strValue, var.bstrVal, nStrLen);
+	#endif
+			strValue[nStrLen - 1] = TEXT('\0');
+
+			break;*/
+
+	default: result = std::string("Default Value");//std::cout << "No value for property" << std::endl;
```

Comment:
Are you sure default value make any sense in this context? It not, it should be error situation

Reply:
yes, it should be an error, but there are some type such as "string" I have not done yet, so I left it as a default value. Further processing will be handled. This should throw an exception